### PR TITLE
Unused enumerated rule improvements

### PR DESCRIPTION
**Pull request by mildm8nnered**
Addresses #5470 

Extends `unused_enumerated` rule to cover cases like

```
list.enumerated().map { idx, _ in idx }
list.enumerated().map { _, elem in elem }
```

and

```
list.enumerated().forEach { print($0) }
list.enumerated().map { $1 }
```

Non-triggering cases are:

```
list.enumerated().map { idx, elem in "\(idx): \(elem)" }
list.enumerated().map { $0 + $1 }
list.enumerated().something().map { _, elem in elem }
```

Of the newly introduced OSS violations, some are false positives. In cases like 

```
   let selection = button.itemIcons.enumerated()
       .filter { _, icon in point.x > rectForItem(icon)!.origin.x }
       .map { index, _ in index }
       .last ?? 0
```

We will only inspect the first closure, and have no way of telling if arguments are accessed in later closures.

### CODE REVIEW

**Comment by SimplyDanny** [comment]
> They are all variations of the following idiom:
> 
> ```
>    let selection = button.itemIcons.enumerated()
>        .filter { _, icon in point.x > rectForItem(icon)!.origin.x }
>        .map { index, _ in index }
>        .last ?? 0
> ```

Well, that's a case I didn't think of. This cannot be recognized reliably, I fear. Not sure the extension of the existing rule is useful then. 🤔

**Comment by mildm8nnered** [comment]
> > They are all variations of the following idiom:
> > ```
> >    let selection = button.itemIcons.enumerated()
> >        .filter { _, icon in point.x > rectForItem(icon)!.origin.x }
> >        .map { index, _ in index }
> >        .last ?? 0
> > ```
> 
> Well, that's a case I didn't think of. This cannot be recognized reliably, I fear. Not sure the extension of the existing rule is useful then. 🤔
> 
> A way around would be to check subsequent `.map` and `.forEach` calls only. Anything more that makes sense?

So I don't think it blows it out of the water necessarily.

Two false alarms on the SwiftLint codebase and 5/17 hits on the OSS - ok, that's not great, but I suspect the problematic idiom is not that common.

We could make this behaviour configurable perhaps ...

**Comment by SimplyDanny** [comment]
> Two false alarms on the SwiftLint codebase and 5/17 hits on the OSS - ok, that's not great, but I suspect the problematic idiom is not that common.

I still think, the rule should only check a few specific cases like `.map`, `.forEach`, ... then. Alternatively, it could stop analysis at `.filter` already.

**Comment by mildm8nnered** [comment]
So I found quite a subtle bug around the 

```
        return Dictionary(
            uniqueKeysWithValues: contents.bridge()
                .components(separatedBy: "\n") // swiftlint:disable:next unused_enumerated
                .map { Int64($0.bridge().lengthOfBytes(using: .utf8)) }
                .reduce(into: [0]) { result, length in
                    let newLineCharacterLength = Int64(1)
                    let lineLength = length + newLineCharacterLength
                    result.append(contentsOf: [(result.last ?? 0) + lineLength])
                }
                .enumerated()
                .map { ($0.offset, $0.element) }
        )
```

case from SwiftLint's own code. We should be detecting a violation here - `.map { ($0.offset, $0.element) }` instead of above.

**Comment by SimplyDanny** [comment]
The fixed violation is not correct, is it?

**Comment by SimplyDanny** [comment]
> Secondly, we were not being very discriminating about which closures we ended up looking at - now we will only inspect the trailingClosure of the call immediately after the enumeration.

I think that the rule should stop on `.filter`. There might always come something after the filter that uses the other value returned by `.enumerated()` even though `.filter` doesn't.

**Comment by mildm8nnered** [comment]
> The fixed violation is not correct, is it?

`list.map { ($0.offset, $0.element) }` this one? 

That should be non-triggering, as the index and element can be accessed via `offset` and `element` on the tuple values that`enumerated()` returns (which I did not know before).

**Comment by mildm8nnered** [comment]
> [The last and only fix](https://github.com/wordpress-mobile/WordPress-iOS/blob/2537c045b9854bea324d159a3ac25815ee083656/WordPress/Classes/Stores/StatsPeriodStore.swift#L1432:30) in the OSS report a mean.

Ah - I didn't spot that - yes, that looks problematic :-( I'll see if I can work out what I've broken.

**Comment by mildm8nnered** [comment]
> > [The last and only fix](https://github.com/wordpress-mobile/WordPress-iOS/blob/2537c045b9854bea324d159a3ac25815ee083656/WordPress/Classes/Stores/StatsPeriodStore.swift#L1432:30) in the OSS report a mean.

That was quite a lucky spot - resolved by https://github.com/realm/SwiftLint/pull/5498/commits/32c445bf939b25d2641f607bb3eef92639622b67 - we were skipping closures that didn't match our trailing closure.

**Comment by mildm8nnered** [comment]
> > Secondly, we were not being very discriminating about which closures we ended up looking at - now we will only inspect the trailingClosure of the call immediately after the enumeration.
> 
> I think that the rule should stop on `.filter`. There might always come something after the filter that uses the other value returned by `.enumerated()` even though `.filter` doesn't.

So I believe that is what the code now does, both for cases like

```
            itemIcons.enumerated()
                   .filter { ↓_, icon in icon == x }
                   .map { index, _ in index }
                   .last
```

```
            itemIcons.enumerated()
                   .filter { ↓$0 == x }
                   .map { $1 }
                   .last
```

We'll only inspect the first closure after the enumerated node in both cases.

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/DuplicateImportsRule.swift [hunk 1]
```diff
@@ -153,7 +153,7 @@ private extension DuplicateImportsRule {
 
         override func visit(_ node: CodeBlockItemListSyntax) -> CodeBlockItemListSyntax {
             let itemsToRemove = node
-                .enumerated()
+                .enumerated() // swiftlint:disable:next unused_enumerated
```
Comment (SimplyDanny) [inline Source/SwiftLintBuiltInRules/Rules/Idiomatic/DuplicateImportsRule.swift:156]: I still wonder why this needs to be skipped. I think that the rule should either stop immediately when encountering a `filter` method or it should skip it and continue checking the subsequent `map`-like methods. If there aren't any of the latter, the rule remains silent.

Comment (mildm8nnered) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/DuplicateImportsRule.swift:156]: Ah - my misunderstanding - I would love to continue checking, but I couldn't find an elegant way to do that - I'll look again.

Or we can just stop on filter and ignore (as opposed to stopping after filter, which is what we're actually doing in the current code here).

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift [hunk 2]
```diff
@@ -44,17 +56,87 @@ private extension UnusedEnumeratedRule {
...
+            addViolation(
+                zeroPosition: firstTokenIsUnderscore ? firstElement.positionAfterSkippingLeadingTrivia : nil,
+                onePosition: firstTokenIsUnderscore ? nil : secondElement.positionAfterSkippingLeadingTrivia
+            )
+        }
+
+        override func visit(_ node: FunctionCallExprSyntax) -> SyntaxVisitorContinueKind {
+            guard node.isEnumerated,
+                  let trailingClosure = node.parent?.parent?.as(FunctionCallExprSyntax.self)?.trailingClosure,
+                  node.parent?.as(MemberAccessExprSyntax.self)?.declName.baseName.text != "filter"
```
Comment (SimplyDanny) [inline Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift:68]: Add `let parent = node.parent` and check the last condition first.

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift [hunk 3]
```diff
@@ -44,17 +56,87 @@ private extension UnusedEnumeratedRule {
...
+                return .visitChildren
+            }
+
+            if let parameterClause = trailingClosure.signature?.parameterClause {
+                guard let parameterClause = parameterClause.as(ClosureShorthandParameterListSyntax.self),
+                      parameterClause.count == 2,
+                      let firstElement = parameterClause.firstParameter,
+                      let secondElement = parameterClause.lastParameter,
+                      case let firstTokenIsUnderscore = firstElement.isUnderscore,
+                      case let lastTokenIsUnderscore = secondElement.isUnderscore,
```
Comment (SimplyDanny) [inline Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift:79]: I don't like the `case let` so much. Perhaps split the `guard` condition.

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift [hunk 4]
```diff
@@ -44,17 +56,87 @@ private extension UnusedEnumeratedRule {
...
+                  let trailingClosure = node.parent?.parent?.as(FunctionCallExprSyntax.self)?.trailingClosure,
+                  node.parent?.as(MemberAccessExprSyntax.self)?.declName.baseName.text != "filter"
+            else {
+                return .visitChildren
+            }
+
+            if let parameterClause = trailingClosure.signature?.parameterClause {
+                guard let parameterClause = parameterClause.as(ClosureShorthandParameterListSyntax.self),
+                      parameterClause.count == 2,
+                      let firstElement = parameterClause.firstParameter,
```
Comment (SimplyDanny) [inline Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift:76]:
```suggestion
                      let firstElement = parameterClause.first,
```

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift [hunk 5]
```diff
@@ -44,17 +56,87 @@ private extension UnusedEnumeratedRule {
...
+                      let secondElement = parameterClause.lastParameter,
```
Comment (SimplyDanny) [inline Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift:77]:
```suggestion
                      let secondElement = parameterClause.last,
```

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift [hunk 6]
```diff
@@ -83,3 +165,19 @@ private extension TuplePatternElementSyntax {
...
+
+private extension ClosureShorthandParameterListSyntax {
+    var firstParameter: ClosureShorthandParameterSyntax? {
+        children(viewMode: .sourceAccurate).first?.as(ClosureShorthandParameterSyntax.self)
+    }
+
+    var lastParameter: ClosureShorthandParameterSyntax? {
+        children(viewMode: .sourceAccurate).last?.as(ClosureShorthandParameterSyntax.self)
+    }
+}
```
Comment (SimplyDanny) [inline Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift:183]: That's just `.first` and `.last`. 😉 
```suggestion
```

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift [hunk 7]
```diff
@@ -44,17 +56,87 @@ private extension UnusedEnumeratedRule {
...
             } else {
-                position = secondElement.positionAfterSkippingLeadingTrivia
-                reason = "When the item is not used, `.indices` should be used instead of `.enumerated()`"
+                self.trailingClosure = trailingClosure
+            }
+
+            return .visitChildren
+        }
+
+        override func visitPost(_ node: ClosureExprSyntax) {
```
Comment (SimplyDanny) [inline Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift:96]: This not going to work for closures in closures. I recommend to go with a separate visitor only collecting appearances of `$0`, `$0.element`, ...

Please also add tests.

**Review by SimplyDanny — CHANGES_REQUESTED** [review]
This is the most the rule can check, I assume. Looks good now! 👍

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift (hunk 6 above)
Comment (mildm8nnered) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift:183]: 🤦

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift (hunk 7 above)
Comment (mildm8nnered) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift:104]: Cool - will take a bit of time to work this out

Comment (mildm8nnered) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift:104]: Do you have an example that you think will fail @SimplyDanny ?

Comment (SimplyDanny) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift:104]: Without trying:

```swift
list.enumerated().map {
    $1.enumerated().forEach { print($0, $1) }
    return $0
}
```

I assume this would trigger the rule on the outer `enumerated()`, even though it's actually validly used.

Comment (mildm8nnered) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift:104]: ok - so I think I have it working for closures within closures, indefinitely. The basic idea is to keep a stack that tracks   the zero and one position for each closure, and then push and pop the data structures onto the stack as we encounter and depart each closure.

It feels to me like it adds a fair bit of complexity for something that is a bit of an edge case, but it does pass all the tests and some new ones. First pass of code is on a PR targeting this branch at 

https://github.com/mildm8nnered/SwiftLint/pull/11

If you think that generally looks good-ish, I'll tidy it up and merge it into this branch.

Comment (SimplyDanny) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift:104]: I added some comments in the other PR.

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift [hunk 8]
```diff
@@ -18,43 +18,177 @@ struct UnusedEnumeratedRule: Rule {
...
+
+            guard (closure.zeroPosition != nil) != (closure.onePosition != nil) else {
+                return
+            }
+
+            addViolation(zeroPosition: closure.zeroPosition, onePosition: closure.onePosition)
+        }
+
+        override func visitPost(_ node: DeclReferenceExprSyntax) {
+            guard var closure = closures.peek(), node.baseName.text == "$0" || node.baseName.text == "$1" else {
```
Comment (mildm8nnered) [inline Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift:163]: > How does it make sure that the closure is the right one?

So when we entered the closure, in `visit(_ node: ClosureExprSyntax)`, if the closure was the trailing closure of an enumerated function call (whose id is stored in `nextClosureId`) we push it onto the stack, so that is what `closures.peek()` will return.

And when we leave the closure, in `visitPost(_ node: ClosureExprSyntax)`, we pop that closure off the stack, and update state for that enclosing closure.

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift [hunk 9]
```diff
@@ -18,43 +18,177 @@ struct UnusedEnumeratedRule: Rule {
...
             Example("for (↓_, foo) in abc.something().enumerated() { }"),
-            Example("for (idx, ↓_) in bar.enumerated() { }")
+            Example("for (idx, ↓_) in bar.enumerated() { }"),
+            Example("list.enumerated().map { idx, ↓_ in idx }"),
+            Example("list.enumerated().map { ↓_, elem in elem }"),
+            Example("list.enumerated().forEach { print(↓$0) }"),
+            Example("list.enumerated().map { ↓$1 }"),
+            Example("""
+            list.enumerated().map {
+                $1.enumerated().forEach { print(↓$1) }
```
Comment (mildm8nnered) [inline Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift:44]: > It feels a bit strange that the rule triggers on the identifiers (that are not missing) even though the rule is about the `enumerated()` method.

I know what you mean - it's not like the `_` case where it's obvious. I think of it as `$0` (or `$1`) being the thing that indicates there's a problem, as we didn't expect to see only one of `$0` or `$1`.

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift (hunk 7 above)
Comment (mildm8nnered) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift:149]: Thanks! I've either adopted them wholesale, or replied inline on this PR ...

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift (hunk 9 above)
Comment (SimplyDanny) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift:44]: I'd still expect the rule to trigger on the `enumerated()` call to be honest. Changing that shouldn't be a big deal, right?

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift (hunk 8 above)
Comment (SimplyDanny) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift:163]: Consider
```swift
list.enumerated().map {
    $1.forEach { print($0) }
    return $1 
}
```
which is expected to trigger on the last `$1`, right? But does it actually? 🤓

**Review by SimplyDanny — CHANGES_REQUESTED** [review]

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift (hunk 8 above)
Comment (mildm8nnered) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift:163]: Dammit. I think I might need to go back to the drawing board on this one :-(

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift (hunk 9 above)
Comment (mildm8nnered) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift:44]: Yep, that should be possible.

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift (hunk 8 above)
Comment (SimplyDanny) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift:163]: Hint: `ShorthandArgumentRule`

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift (hunk 2 above)
Comment (mildm8nnered) [reply Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift]: So I didn't go the `ShorthandArgumentRule` route. Instead, when I see a closure that is not the one I'm looking for, I  push it onto the stack anyway, and then stop counting `$0`'s and `$1`'s. Seems to work ok for that test case.

**Comment by mildm8nnered** [comment]
So we now place the violation just before the `enumerated()` call for the `$0` and `$1` cases, and `$0` or `$1` should be ignored in irrelevant closures.

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift [hunk 10]
```diff
@@ -18,43 +18,211 @@ struct UnusedEnumeratedRule: Rule {
...
+
+            guard (closure.zeroPosition != nil) != (closure.onePosition != nil) else {
+                return
+            }
+
+            addViolation(
+                zeroPosition: closure.zeroPosition,
+                onePosition: closure.onePosition,
+                enumeratedPosition: closure.enumeratedPosition
+            )
```
Comment (SimplyDanny) [inline Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift:181]: The case that the closure IDs don't match seem impossible. Simplify: 
```suggestion
            if let closure = closures.pop(), (closure.zeroPosition != nil) != (closure.onePosition != nil) {
                addViolation(
                    zeroPosition: closure.zeroPosition,
                    onePosition: closure.onePosition,
                    enumeratedPosition: closure.enumeratedPosition
                )
            }
```

**Review by SimplyDanny — APPROVED** [review]

File: Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift [hunk 11]
```diff
@@ -18,57 +18,221 @@ struct UnusedEnumeratedRule: Rule {
...
+                return $0
+            }
+            """, excludeFromDocumentation: true)
+            ,
+            Example("""
+            list.↓enumerated().map {
+                $1.forEach { print($0) }
+                return $1
+            }
+            """, excludeFromDocumentation: true)
```
Comment (SimplyDanny) [inline Source/SwiftLintBuiltInRules/Rules/Idiomatic/UnusedEnumeratedRule.swift:74]: Thanks for excluding the examples! 👍

**Comment by SimplyDanny** [comment]
The [last OSS finding](https://github.com/wireapp/wire-ios/blob/5d2f52ccf3ee0afaf32f64c2ce50c01081415284/wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/ViewModel/ConversationListViewModel.swift#L551:22) is a false positive. The implementation needs to check likewise for `$0.offset` in addition to `$0.element`.

**Comment by mildm8nnered** [comment]
> The [last OSS finding](https://github.com/wireapp/wire-ios/blob/5d2f52ccf3ee0afaf32f64c2ce50c01081415284/wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/ViewModel/ConversationListViewModel.swift#L551:22) is a false positive. The implementation needs to check likewise for `$0.offset` in addition to `$0.element`.

So I think that is actually a hit:

```
            sections.enumerated().forEach {
                delegate?.listViewModel(self, didUpdateSection: $0.offset)
            }
```

`$0.element` isn't referenced, so this could be

```
            sections.indices.forEach {
                delegate?.listViewModel(self, didUpdateSection: $0)
            }
```

I was providing the wrong reason for the `$0` and `offset/element` cases though 🤦 which I've fix in the last commit.

**Comment by SimplyDanny** [comment]
> > The [last OSS finding](https://github.com/wireapp/wire-ios/blob/5d2f52ccf3ee0afaf32f64c2ce50c01081415284/wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/ViewModel/ConversationListViewModel.swift#L551:22) is a false positive. The implementation needs to check likewise for `$0.offset` in addition to `$0.element`.
> 
> 
> 
> So I think that is actually a hit:
> 
> 
> 
> ```
> 
>             sections.enumerated().forEach {
> 
>                 delegate?.listViewModel(self, didUpdateSection: $0.offset)
> 
>             }
> 
> ```
> 
> 
> 
> `$0.element` isn't referenced, so this could be
> 
> 
> 
> ```
> 
>             sections.indices.forEach {
> 
>                 delegate?.listViewModel(self, didUpdateSection: $0)
> 
>             }
> 
> ```

You're right. 👍 The implementation assumes that any member access on `$0` is either `element` or `offset`. Seems correct, but wasn't instantly clear to me.

**Comment by mildm8nnered** [comment]
> You're right. 👍 The implementation assumes that any member access on `$0` is either `element` or `offset`. Seems correct, but wasn't instantly clear to me.

So this logic here should handle that:


```
              if node.baseName.text == "$0" {
                    if node.parent?.as(MemberAccessExprSyntax.self)?.declName.baseName.text == "element" {
                        $0.onePosition = node.positionAfterSkippingLeadingTrivia
                    } else {
                        $0.zeroPosition = node.positionAfterSkippingLeadingTrivia
                    }
                } else {
                    $0.onePosition = node.positionAfterSkippingLeadingTrivia
                }
            }
```

`$0.element` we effective treat as though we saw `$1` ...

**Comment by SimplyDanny** [comment]
> You're right. 👍 The implementation assumes that any member access on `$0` is either `element` or `offset`. Seems correct, but wasn't instantly clear to me.

Well, actually `$0.offset` can also be addressed by `$0.0` or even crazy things like `$0.self.0` or `$0.self.self.0` or ... While the last examples seem abnormal, I've seen `$0.0` and `$0.1` quite often.

**Comment by mildm8nnered** [comment]
> > You're right. 👍 The implementation assumes that any member access on `$0` is either `element` or `offset`. Seems correct, but wasn't instantly clear to me.
> 
> Well, actually `$0.offset` can also be addressed by `$0.0` or even crazy things like `$0.self.0` or `$0.self.self.0` or ... While the last examples seem abnormal, I've seen `$0.0` and `$0.1` quite often.

We could probably cover `$0.1` in the same way that we do `element` if you think we should ...

**Comment by SimplyDanny** [comment]
> > > You're right. 👍 The implementation assumes that any member access on `$0` is either `element` or `offset`. Seems correct, but wasn't instantly clear to me.
> > 
> > 
> > Well, actually `$0.offset` can also be addressed by `$0.0` or even crazy things like `$0.self.0` or `$0.self.self.0` or ... While the last examples seem abnormal, I've seen `$0.0` and `$0.1` quite often.
> 
> We could probably cover `$0.1` in the same way that we do `element` if you think we should ...

Doesn't seem to be a big deal. I think we should.

**Comment by mildm8nnered** [comment]
> Doesn't seem to be a big deal. I think we should.

Cool - should be good now ...
